### ISSUE TITLE
Creating a Java VM fails on Android 7.1.

### ISSUE THREAD

**Issue by feicong** [CalebFenton/native-harness-target#1]
i try to create java vm on android 7.1 with this code. 
```
void *librt_dso = dlopen("libart.so", RTLD_NOW);
void *libandroid_runtime_dso = dlopen("libandroid_runtime.so", RTLD_NOW);
```

the result of the operation was program directly abort. i try to find `JNI_CreateJavaVM()` use this code:
```
JNI_CreateJavaVM_t JNI_CreateJavaVM;
    JNI_CreateJavaVM = (JNI_CreateJavaVM_t) dlsym(RTLD_DEFAULT, "JNI_CreateJavaVM");
    if (!JNI_CreateJavaVM) {
        return -2;
    }
```

still returns null. 

is there any other solution?

**Comment by neogranzeon** [comment]
1. For error -2, the main problem is Signal Chain library on ART. You will need to export JNI function related to the library and set dynamic export to LDFLAGS. With this, you can run the ART-based JavaVM on some early ART version like Lollipop (confirmed).

2. On Android 7.1, since the "core/java/com/android/internal/util/WithFramework.java" has been deleted [1] you will get error -3.

Reference:
[1]. https://android-review.googlesource.com/#/c/platform/frameworks/base/+/157981/

**Comment by CalebFenton** [comment]
This might've been fixed here https://github.com/CalebFenton/native-harness-target/pull/2

**Comment by spiiin** [comment]
My question is a little off topic, but there are no places to discuss it. I built executable and test it with adb - it works fine, jvm was created. Next, I want to use this code as a library for loading it via android python to create JavaVM and JNIenv inside it in interactive mode. I used ctypes for it:
```python
from ctypes import cdll
a = cdll.LoadLibrary("/data/myapps/main.out") #path to binary, it works via terminal

import ctypes
class JavaVM(ctypes.Structure):
    _fields_ = [("functions", ctypes.POINTER(ctypes.c_void_p))]
    
class JniENV(ctypes.Structure):
    _fields_ = [("functions", ctypes.POINTER(ctypes.c_void_p))]
    
jvm = ctypes.POINTER(JavaVM)()
env = ctypes.POINTER(JniENV)()

a.init_jvm(ctypes.byref(jvm),ctypes.byref(env))
```
I checked ctypes signatures with version of module with empty init_jvm function, seems it correct. But when I called this function with real code, it fails, and in logcat I see:
```
2018-12-25 15:54:04.123 24658-24658/? E/libsigchain: InitializeSignalChain is not exported by the main executable.
    
    --------- beginning of crash
2018-12-25 15:54:04.124 24658-24658/? A/libc: Fatal signal 6 (SIGABRT), code -6 in tid 24658 (python3.6)
2018-12-25 15:54:04.131 24979-24979/? I/debuggerd: type=1400 audit(0.0:649): avc: denied { read } for name="python3.6" dev=mmcblk0p8 ino=34307 scontext=u:r:debuggerd:s0 tcontext=u:object_r:app_data_file:s0:c512,c768 tclass=file permissive=1
2018-12-25 15:54:04.131 24979-24979/? I/debuggerd: type=1400 audit(0.0:650): avc: denied { open } for name="python3.6" dev=mmcblk0p8 ino=34307 scontext=u:r:debuggerd:s0 tcontext=u:object_r:app_data_file:s0:c512,c768 tclass=file permissive=1
2018-12-25 15:54:04.131 24979-24979/? I/debuggerd: type=1400 audit(0.0:651): avc: denied { getattr } for path="/data/data/ru.iiec.pydroid3/files/arm-linux-androideabi/bin/python3.6" dev=mmcblk0p8 ino=34307 scontext=u:r:debuggerd:s0 tcontext=u:object_r:app_data_file:s0:c512,c768 tclass=file permissive=1
2018-12-25 15:54:04.151 24658-24658/? A/libc: failed to resend signal during crash: Operation not permitted
2018-12-25 15:54:10.211 2571-3994/com.google.android.gms I/CastDatabase: Saving the database
2018-12-25 15:54:10.230 2571-3994/com.google.android.gms I/SQLiteCastStore: [CastNetworkInfo] saved: 0, skipped: 0, [CastDeviceInfo]: saved 0, skipped 0, Paired Guest Mode DeviceInfo] saved: 0, skipped: 0. [ProbedNetworks]: saved 0, skipped: 0, [ProbedSocketAddress]: saved 0, skipped: 0
```
Does it mean, that libsigchain functions must exist inside python3.6 application, not only inside library, that created JavaVM? If so, any chance to create those functions when application already started?
I used 3rd-party android python interpreter, so can't add them to the source code.

My environment:
Android 7.1.2
Python interpreter https://play.google.com/store/apps/details?id=ru.iiec.pydroid3&hl=ru
Sorry for my English, it's not my native.

**Comment by CalebFenton** [comment]
@spiiin you should consider joining rednaga.slack.com and asking someone there. This project has been incorporated into https://github.com/rednaga/native-shim.

Two things you can try: 1.) add debugging to the vm.c code so you see how far it gets. 2.) try loading a simple native library and calling a simple function with no arguments. I'm not an expert with ctypes, but some of the double pointer stuff, how you declare the classes, and how you create the jvm and env pointers seems a little weird.

**Comment by spiiin** [comment]
Thanks for the answer. I solved my case by just setting LD_LIBRARY_PATH environment variable. I copied  libart.so and created my own copy libsigchain.so, put the code of init_vm and stubs functions in it, and load it via ctypes from python. It works. ctypes declarations is correct. If someone needs it, an example is here:
https://github.com/brunoalfirevic/VimClojureInterface/blob/master/python/pyjni.py
